### PR TITLE
Rework hook registration

### PR DESCRIPTION
**Pull request by nickrum**
**Notice**: This PR tweaks the structure of how hooks are created and updates some of the event names. Please update your custom hooks if you were relying on any.

---

New syntax for hook registration:

```js
export default ({ filter, action, init, schedule }) => {
  filter('items.create', () => {
    console.log('Creating Item!');
  });

  action('items.create', () => {
    console.log('Item created!');
  });

  init('app.before', () => {
    console.log('Hello there!');
  });

  schedule('*/5 * * * * *', () => {
    console.log('Surprise!');
  });
};
```

Thanks to @smilledge who came up with the initial idea [here](https://github.com/directus/directus/discussions/7744#discussioncomment-1288090) and [here](https://github.com/directus/directus/discussions/7843).

### CODE REVIEW

File: api/src/services/fields.ts [hunk 1]
```diff
@@ -425,7 +425,7 @@ export class FieldsService {
 			await this.schemaCache.clear();
 		}
 
-		emitAsyncSafe(`fields.delete`, {
+		emitter.actionSafe(`fields.delete`, {
```
Comment (rijkvanzanten) [inline api/src/services/fields.ts:428]: Every action is "safe" right? Let's rename it "action" for consistencies sake

**Comment by rijkvanzanten** [comment]
What's left to do on this PR to make it ready for review?

**Comment by nickrum** [comment]
The _safe_ vs _not safe_ thing you already mentioned, a few hooks that were filters but do not really make sense as filters, updating the extension templates and docs.

**Comment by smilledge** [comment]
Some thoughts/questions 🧐

## What is the correct way to modify the payload of a filter?

**1. Return the updated value (or the same value if unchanged)**
```javascript
filter('items.create', (payload) => {
    return { ...payload: panda: '🐼' }
});
```
This has a few issues if you have multiple listeners:
- What order are the listeners called in?
- How are the returned values merged? (currently there seems to be a mix of merging and using the last returned value)
- If you return the initial value will it override changes from another filter?
- What happens if the return value is `undefined`?

**2. Mutate the object without returning anything, eg;**
```javascript
filter('items.create', (payload) => {
    payload.panda = '🐼'
});
```
This has some issues too:
 - It can only work with objects (you could never filter a string for example)
 - All filters currently pass the payload by reference (and use the payload object after the filter is applied)
 - This could still have inconsistent output depending of the listener order if you are modifying the same payload field

## Should filters have a priority / order option?
```javascript
// const payload = { panda: '🦆' };

filter('items.create', (payload) => {
    if (payload.panda === '🦆') {
        return { ...payload: panda: '🐼' };
    }
    return payload;
});

filter('items.create', (payload) => {
    return { ...payload: panda: '🦥' };
});

filter('items.create', (payload) => {
    return payload;
});
```
In this example, what would the value of `panda` be? It seems like it would be possible to get  a 🦥, 🐼 or 🦆 depending on the order the listeners were called in.

If there was an option to define a filter priority they could be run sequentially with the result of the previous listener passed to the next in the list. This would ensure you get a consistent result.

**Comment by rijkvanzanten** [comment]
Great questions @smilledge!

> What is the correct way to modify the payload of a filter?

I _think_ I would personally prefer the "return" method. It means that it's a little more obvious and explicit what's being returned in the hook. It also allows an easier "If X, return Y" flow in that approach, and has the benefit (like you stated) that it's not locked to object manipulation.

> Should filters have a priority / order option?

As of right now, the filters would be registered in order that function is called, which means that the payload is processed in order of activation of those hooks. Adding a priority parameter would allow the user to more explicitly control this flow, but also means that it's a little more confusing to trace what's going on in what order if you're pulling in hooks from external sources 🤔 

In terms of 80/20, I'm also curious to see how often this situation would happen realistically. I would assume (evil word in OSS) that most users would stick to a limited amount of hooks per event. If the user is the one creating the hooks, the user should be aware of what changes are made in the various hooks, and make sure there's no conflicts. The only pain point i foresee there is if you install a third-party hook through an extension which might do some things to the payload you didn't expect, but that also might be a more of a documentation/expectation management problem than a order/priority problem 🤔

**Comment by smilledge** [comment]
> I _think_ I would personally prefer the "return" method

Yeah I agree :+1: 

> I would assume (evil word in OSS) that most users would stick to a limited amount of hooks per event

Not sure I'd agree with this. Even without third party extensions I would assume (😅) that people will want to modularize their codebase by feature. Which will likely mean using the same event multiple times.

# Suggested implementation 
I put together a quick prototype for how I think filters should be implemented. Hopefully this helps to explain some of the concerns I have with the current implementation.

## Filter rules

- Filters will receive the payload as the first argument
- Filters should return a value of the same type as they received
- Additional information can be passed in the second argument (this should not be returned or modified by the filter)
- Filter payloads can be of any data type

## Filter process
Filters are run one-by-one and the result of the previous filter is passed as the payload to the next. The returned values are no longer merged at the end; instead the filters return the updated value.

Example implementation: https://github.com/smilledge/directus/commit/55b5ee893e21c4e04f30d1b41833ab55307ec245

**Current implementation issues:**

You can accidentally override the result of another filter:
```javascript
// Payload: { title: 'Title', body: 'Body' }
filter('items.create', ({ payload }) => payload);
filter('items.create', ({ payload }) => ({ ...payload, title: `${payload.title} 🐼`, body: '🍔' }));
filter('items.create', ({ payload }) => payload);
// Result: "{ title: 'Title', body: 'Body' }"
```

You can't modify the same value multiple times:
```javascript
// Payload: { title: 'Title', body: 'Body' }
filter('items.create', ({ payload }) => ({ ...payload, title: `${payload.title} 🐼` }));
filter('items.create', ({ payload }) => ({ ...payload, title: `${payload.title} 🦆` }));
filter('items.create', ({ payload }) => ({ ...payload, title: `${payload.title} 🦝`, body: '🍔' }));
// Result: "{ title: 'Title 🐼', body: 'Body' }"
```

Also, you can't really filter non-object data types if you have more than one filter listening to an event.

**Suggested implementation results:**

```javascript
// Payload: { title: 'Title', body: 'Body' }
filter('items.create', (payload) => payload);
filter('items.create', (payload) => ({ ...payload, title: `${payload.title} 🐼`, body: '🍔' }));
filter('items.create', (payload) => payload);
// Result: "{ title: 'Title 🐼', body: '🍔' }"
```

```javascript
// Payload: { title: 'Title', body: 'Body' }
filter('items.create', (payload) => ({ ...payload, title: `${payload.title} 🐼` }));
filter('items.create', (payload) => ({ ...payload, title: `${payload.title} 🦆` }));
filter('items.create', (payload) => ({ ...payload, title: `${payload.title} 🦝`, body: '🍔' }));
// Result: "{ title: 'Title 🐼 🦆 🦝', body: '🍔' }"
```

```javascript
// Payload: { title: "Title" }
filter('items.create', ({ payload }) => undefined);
filter('items.create', ({ payload }) => ({ ...payload, title: `${payload.title} 🐼` }));
filter('items.create', ({ payload }) => payload);
filter('items.create', ({ payload }) => ({ ...payload, title: `${payload.title} 🦆`, body: '🍔' }));
filter('items.create', ({ payload }) => ({ ...payload, title: `${payload.title} 🦝` }));
filter('items.create', ({ payload }) => payload);
// Result: "{ title: 'Title 🐼 🦆 🦝', body: '🍔' }"
```

IMO these are the results you would expect?

## Filter ordering
Filters can optionally be registered with an order (any positive or negative number). The filters will then be run in ascending order.

Example implementation: https://github.com/smilledge/directus/commit/29587738df54ee8e13b4fec7d6d7cf7512024375 

**Usage:**
```javascript
// Payload: { title: "Title" }
filter('items.create', (payload) => ({ ...payload, title: `${payload.title} 🐼` }), { order: 10 });
filter('items.create', (payload) => ({ ...payload, title: `${payload.title} 🦆` }));
filter('items.create', (payload) => ({ ...payload, title: `${payload.title} 🦝` }), { order: -1 });
// Result: "{ title: "Title 🦝 🦆 🐼" }"
```

-----

Hopefully that make more sense? Thoughts/concerns?

**Comment by rijkvanzanten** [comment]
> The returned values are no longer merged at the end; instead the filters return the updated value.

This makes total sense to me

> IMO these are the results you would expect?

I agree 👍🏻 

The only thing I'm worried about is the confusion around "order" 🤔 It might make it a little harder to figure out what order things are running in, and I'm slightly worried people will just set it to "Infinity" or "-1" to force it at the start/end (I doubt extension makers will use an arbitrary value to put it in the "middle". What happens if two filters have the same order value?

**Comment by smilledge** [comment]
Yeah order probably isn't that important at the moment and might not be worth implementing 🤷‍♂️

> The only thing I'm worried about is the confusion around "order" 

It would be an optional field and I'm assuming the majority of the time you wouldn't set an order.

However, if you get into a situation were you need your filter to run before/after a third party extension you currently have no way of doing this. 

This isn't really an issue when you're only using local extensions. But it might limit your ability to modularize your extensions the way you want to.

> What happens if two filters have the same order value?

They would just run in the ordered they were registered.

**Comment by Attacler** [comment]
This looks cool and way better!
Altho i'm wondering if this would break current hooks or how should i see this.
Also, will this add a way to dynamicly remove & add hooks?

**Comment by rijkvanzanten** [comment]
> Altho i'm wondering if this would break current hooks or how should i see this.

Yes it will. It is a breaking change, though the migration steps to the new hook structure should be trivial. That's a necessary evil of being in RC to get this out the door before v9.0.0 release.

> Also, will this add a way to dynamicly remove & add hooks?

Dynamically adding should be possible. There's no "remove event listener" capability included in this at this point

File: docs/guides/api-hooks.md [hunk 2]
```diff
@@ -14,7 +14,39 @@ Custom hooks are dynamically loaded from within your extensions folder. By defau
...
+
+There are four types of hooks to choose from.
+
+### Filter
+
+A filter executes at a certain event and recieves some data that can be altered.
+
+### Action
+
+An action executes after a certain event and might recieve some data related to the event.
```
Comment (paescuj) [inline docs/guides/api-hooks.md:27]:
```suggestion
An action executes after a certain event and might receive some data related to the event.
```

File: docs/guides/api-hooks.md [hunk 3]
```diff
@@ -14,7 +14,39 @@ Custom hooks are dynamically loaded from within your extensions folder. By defau
...
 ```
 
-## 2. Define the Event
+## 2. Determine the hook type
+
+There are four types of hooks to choose from.
+
+### Filter
+
+A filter executes at a certain event and recieves some data that can be altered.
```
Comment (paescuj) [inline docs/guides/api-hooks.md:23]:
```suggestion
 A filter executes at a certain event and receives some data that can be altered.
 ```

File: docs/guides/api-hooks.md [hunk 4]
```diff
@@ -14,7 +14,39 @@ Custom hooks are dynamically loaded from within your extensions folder. By defau
 /extensions/hooks/<hook-id>/index.js
 ```
 
-## 2. Define the Event
+## 2. Determine the hook type
```
Comment (paescuj) [inline docs/guides/api-hooks.md:17]:
```suggestion
## 2. Determine the Hook Type
```

**Comment by Oreilles** [comment]
This would allow both using `items.create` and `<my_custom_collection>.items.create`, the former being executed before the latter:
```ts
// api/src/emitter.ts
public getFilterListeners(event: string, meta: Record<string, any>): FilterHandler[] {
	const listeners = this.filterEmitter.listeners(event);
	if (event.startsWith('items')) {
		return listeners.concat(this.filterEmitter.listeners(`${meta.collection}.${event}`));
	}
	return listeners;
}

public async emitFilter<T>(event: string, payload: T, meta: Record<string, any>, context: HookContext): Promise<T> {
	const listeners = this.getFilterListeners(event, meta);
```

This is just a proposal, I feel like it would offer nice and cheap syntactic sugar because item filters are often collection dependant - but let's see what you @rijkvanzanten @nickrum think. Not a hill I'm willing to die on 🌱. This PR's already looking good to me.

**Comment by rijkvanzanten** [comment]
Like it! Way cleaner than forcing the reliance on `*`, which comes with its own drawbacks 🙂

**Comment by Oreilles** [comment]
Something I noticed: `emitAction` and `emitInit` catch and logs error but do not rethrow them, so they're essentially silenced ? Also, `emitFilter` do not catch/log errors. 
Is that expected ?

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**